File: cpp/739_daily_temperatures.cpp
#include <stack>
#include <vector>
using namespace std;

class Solution {
 public:
  vector<int> dailyTemperatures(vector<int>& temperatures) {
    stack<int> stack;
    int n = temperatures.size();
    vector<int> result(n, 0);
    for (int i = 0; i < n; ++i) {
      while (!stack.empty() && temperatures[stack.top()] < temperatures[i]) {
        int day = stack.top();
        stack.pop();
        result[day] = i - day;
      }
      stack.push(i);
    }
    return result;
  }
};

File: rust/739_daily_temperatures.rs
impl Solution {
    pub fn daily_temperatures(temperatures: Vec<i32>) -> Vec<i32> {
        let mut stack = vec![];
        let n = temperatures.len();
        let mut result = vec![0; n];

        for i in 0..n {
            while let Some(&day) = stack.last() {
                if temperatures[day] < temperatures[i] {
                    stack.pop();
                    result[day] = (i - day) as i32;
                } else {
                    break;
                }
            }
            stack.push(i);
        }

        result
    }
}
